jumanji/environments/logic/sliding_tile_puzzle/generator.py

```diff
@@ -0,0 +1,145 @@
+# Copyright 2022 InstaDeep Ltd. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import abc
+from typing import Tuple
+
+import chex
+import jax
+from jax import numpy as jnp
+
+
+class Generator(abc.ABC):
+    @property
+    @abc.abstractmethod
+    def grid_size(self) -> int:
+        """Size of the puzzle (n x n grid)."""
+
+    @abc.abstractmethod
+    def __call__(self, key: chex.PRNGKey) -> Tuple[chex.Array, Tuple[int, int]]:
+        """Generate a problem instance.
+
+        Args:
+            key: jax random key for any stochasticity used in the instance generation process.
+
+        Returns:
+            A tuple of a 2D array representing a problem instance and a tuple
+            indicating the position of the empty tile.
+        """
+
+
+class RandomGenerator(Generator):
+    """A generator for random Sliding Tile Puzzle configurations."""
+
+    def __init__(self, grid_size: int):
+        """Initialize the RandomGenerator.
+
+        Args:
+            grid_size: The size of the puzzle (n x n grid).
+        """
+        self._grid_size = grid_size
+
+    @property
+    def grid_size(self) -> int:
+        return self._grid_size
+
+    def __call__(self, key: chex.PRNGKey) -> Tuple[chex.Array, chex.Array]:
+        """Generate a random Sliding Tile Puzzle configuration.
+
+        Args:
+            key: PRNGKey used for stochasticity in the generation process.
+
+        Returns:
+            A tuple of a 2D array representing a problem instance and a tuple
+            indicating the position of the empty tile.
+        """
+        # Create a list of all tiles
+        tiles = jnp.arange(self._grid_size * self._grid_size)
+
+        # Shuffle the tiles
+        key, subkey = jax.random.split(key)
+        shuffled_tiles = jax.random.permutation(subkey, tiles)
+
+        # Reshape the tiles into a 2D array
+        puzzle = jnp.reshape(shuffled_tiles, (self._grid_size, self._grid_size))
+
+        # Find the position of the empty tile
+        empty_tile_position = jnp.stack(
+            jnp.unravel_index(jnp.argmax(puzzle == 0), puzzle.shape)
```

Comment:
Seeing as you use the empty tile here also I think make a constants file and just put that in so you can import it in both places easily